### ISSUE TITLE
Parameter used in dynamic dispatch marked as unused

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#518]
```php
function reindex(array $arr, string $methodName): array
{
    $ret = [];

    foreach ($arr as $element) {
        $ret[$element->$methodName()] = true;
    }

    return $ret;
}
```
psalm with --find-dead-code should not complain about `$methodName` not being used in this function.

**Comment by weirdan** [comment]
Though I just noticed there's a bug (method called on the wrong ref)

**Comment by muglug** [comment]
> This is probably another problem psalm could detect: calling methods on non-object

Yeah, it fails because those checks only kick in for known methods (vs `$methodName`). I'll fix that.

**Comment by muglug** [comment]
Other issue ticketed here: https://github.com/vimeo/psalm/issues/519

I've updated the code you posted to make it valid.

**Comment by muglug** [comment]
Thanks! Dead code detection has only been thoroughly tested on Psalm's own codebase, so there are likely more false positives. But I'll keep fixing :)
